### ISSUE TITLE
PHP 8.0 | Add support for named function call parameters

### ISSUE THREAD

**Issue by jrfnl** [squizlabs/PHP_CodeSniffer#3159]
## Introduction

PHP 8.0 introduces named function call parameters:
```php
array_fill(start_index: 0, num: 100, value: 50);

// Using reserved keywords as names is allowed.
array_foobar(array: $array, switch: $switch, class: $class);
```

Ref: https://wiki.php.net/rfc/named_params

## Current tokenization

PHP 8.0 natively tokenizes the parameter label as `T_STRING`. This is the same in PHP 7.4.
When a reserved keyword is used as a parameter label, the tokenizer will tokenize it as the token for the reserved keyword.

When run through the PHPCS tokenizer, the parameter label, in most cases, will currently be incorrectly retokenized as `T_GOTO_LABEL`.
The reserved keyword tokenization as the reserved keyword token remains the same in most cases, with a few exceptions where the token is retokenized as `T_STRING`.

The `:` will currently tokenize to `T_COLON` or `T_INLINE_ELSE` in PHPCS, depending on the wider context of the code sample.

## Proposal for handling this in PHPCS

I'm currently looking into adding support for named function call parameters to PHPCS.

Based on my initial findings, I'd like to propose the following:
* Introduce a new PHPCS native token constant `T_PARAM_NAME` (or `T_PARAM_NAME_LABEL` or `T_PARAM_LABEL`) and consistently tokenize the parameter label as such.
* Ensure that the `:` following a parameter label is always tokenized as `T_COLON`.

Reasoning:
* The new `T_PARAM_NAME` token will make it a lot easier to prevent confusion in the ternary tokenizer and can prevent the colon from being misidenitified as `T_INLINE_ELSE`.
* `T_STRING` is already used for a lot of things and often enough it is hard to determine what a `T_STRING` signifies.
    Without the new token, any sniff looking for `T_STRING` would need to be checked and possibly adjusted to take parameter name labels into account when determining if this is a `T_STRING` the sniff should handle.
    With the new token, existing sniffs looking for `T_STRING` would not need any adjustments and would continue to work as they are.
* For sniffs looking for the `T_COLON` token, with the new `T_PARAM_NAME` token, it will be easy to determine the context of the `T_COLON`.

@gsherwood What's your opinion on this proposal ?

## Affected sniffs

I haven't made a full inventory yet of sniffs which will be affected.

At the very least, I imagine any and all sniffs which examine and fix "function call argument spacing" will need to be reviewed and adjusted to allow for parameter name labels.

**Comment by jrfnl** [comment]
Little addition: I only just noticed that the name + the colon for a `goto` label are joined together into the `T_GOTO_LABEL` token. We could make the choice to do the same for the parameter name labels.

**Comment by gsherwood** [comment]
I agree with everything in this proposal. A new `T_PARAM_NAME` token, with the `T_COLON` tokenized separately sounds like the best solution.

I'll have a look into that `T_GOTO` token and throw something in the 4.0 roadmap for it. Thanks for raising that.

**Comment by jrfnl** [comment]
@gsherwood Thanks for the feedback. I'll continue work on this along those lines.

**Comment by jrfnl** [comment]
Okay, so I've finished the work on this. The PR would conflict with the open PR goto fix PR #3177 for the `package.xml` file, but I can pull it anyway and rebase either one depending on which gets merged first.

I've written an abundance of unit tests into redundancy to accompany the PR, but haven't been able to come up with a code sample which would break the pattern I've used to detect named parameters in the PHP tokenizer class, which is basically: _"open parenthesis or comma, followed by a word, followed by a colon (disregarding whitespace and comments)"_.

If anyone can think of a code sample with valid PHP code which uses that pattern, but isn't a named parameter, please leave a comment here.

**Comment by jrfnl** [comment]
> Affected sniffs
> 
> I haven't made a full inventory yet of sniffs which will be affected.
> 
> At the very least, I imagine any and all sniffs which examine and fix "function call argument spacing" will need to be reviewed and adjusted to allow for parameter name labels.

I've ran every build-in standard against the test case file which is added in PR #3178 and I have not been able to find any sniff which will need adjusting for named parameters.

All sniffs still seem to work as expected.

There are three sniffs for which adding some unit tests could be considered, though their behaviour is fine as-is and the unit tests would only go to confirm and safeguard that:
* `Generic.Functions.FunctionCallArgumentSpacing`
* `PEAR.Functions.FunctionCallSignature`
* `PSR2.Functions.FunctionCallSignature`
